### ISSUE TITLE
Add combination indexes 

### ISSUE THREAD

**Issue by jdoliner** [rethinkdb/rethinkdb#1560]
There are 2 parts to this proposal. The first is that I think we should allow for objects as keys both in secondary indexes and primary. The reason being that if you try to create a compound index on a set of fields `["a", "b", "c"]` and some of your documents are missing fields it gets really awkward really quickly. The standard way to create this index is:

``` Python
table.index_create("a_b_c", lambda doc: [doc["a"], doc["b"], doc["c"]])
```

However if any of these are missing the whole result returns an error which means that the row doesn't get at all. To get around this you have to type `default` a lot which is a pain. However if we allowed objects then you could do:

``` Python
table.index_create("a_b_c", lambda doc: doc.pluck("a", "b", "c"))
```

which is very convenient. I've had a few people mention that this would be nice.

This feature gets a lot more useful when you add a `combination=True` flag to `index_create` here's how it would work:

``` Python
table.index_create("a_b_c", r.row.pluck("a", "b", "c"), combination=True)
table.get_all({"a" : 1})
table.get_all({"a" : 1, "b" : 2})
```

basically you get an indexed version of the shorthand filter syntax.

This is also really easy to implement because it uses exactly the same code paths as multi indexes.
